# Project: What's The Hot Topic In Town? - kelvin.ahiakpor & emmanuel.acquaye

This notebook integrates Phase 1-5 of the What's The Hot Topic in Topic Town? project:   
**1. [Web Scraping](#Phase_1)**  
**2. [Display News](#Phase_2)**  
**3. [Social Sentiment Analysis](#Phase_3)**  
**4. [Summarize News Articles](#Phase_4)**  
**5. [Deploy Functional App](#Phase_5)**  

### Information Retrieval

# Phase 1           
Web Scraping

The self-created rubric, in our repository, explains the requirement for a proper execution of this phase as seen below.   
**Description:** Build a Spider to scrape news from at least 1 credible news source (AllAfrica, BBC, The New
York Times, Reuters, Foreign Affairs).  
**Note:** This notebook **only** shows the code and instructions for web scraping with a spider with .py files. Scrapy is built in python and will not run in this environment. To learn how to build a scrapy spider, read the Scrapy documentation. [Scrapy Tutorial](https://docs.scrapy.org/en/latest/intro/tutorial.html)

### Repository Link 

Here is a link to our repository:

[What's The Hot Topic In Town?](https://github.com/kelvin-ahiakpor/Whats.The.Hot.Topic.In.Town)

### Imports

In [1]:
import scrapy
import pycountry
import pandas as pd
import random
import time
import sqlite3 as sql
from scrapy.downloadermiddlewares.useragent import UserAgentMiddleware

**Create a .py file called** `news_scraper.py` **for create Spider classes**   

In [2]:
class TrendsNewsSpider(scrapy.Spider):
    name = 'trends_news'
    
    custom_settings = {
        'DOWNLOAD_DELAY': 2,  # Default delay of 2 seconds
        'RANDOMIZE_DOWNLOAD_DELAY': True,
        'CONCURRENT_REQUESTS': 1,
        'RETRY_ENABLED': True,
        'RETRY_TIMES': 3,
        'DOWNLOADER_MIDDLEWARES': {
            'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware': None,
            'newsspider.middlewares.RandomUserAgentMiddleware': 400,
        }
    }
    
    def __init__(self, *args, **kwargs):
        super(TrendsNewsSpider, self).__init__(*args, **kwargs)
        self.data = []

    def start_requests(self):
        country = getattr(self, 'country', None)
        if country is not None:
            if ' ' in country:
                country1 = country.replace(' ', '-')
                url = 'https://trends24.in/' + country1.lower() + '/'
            else:
                url = 'https://trends24.in/' + country.lower() + '/'
            url_allafrica = f'https://allafrica.com/{country.lower()}/'
            
            yield scrapy.Request(url=url, callback=self.parse_trends, meta={'country': country})
            yield scrapy.Request(url=url_allafrica, callback=self.parse_allafrica, meta={'country': country})
        else:
            self.log("No country provided")

    def parse_trends(self, response):
        country = response.meta['country']
        trend_list = response.css('.list-container .trend-card__list')
        trends = []
        for trend in trend_list[0].css('li a::text').getall():
            trends.append(trend.strip())
        
        self.log(f"Top 50 trends in {country}:")
        for trend in trends[:50]:
            self.log(trend)
            #time.sleep(random.uniform(1, 3))  # Random delay between requests
    
    def parse_allafrica(self, response):
        country = response.meta['country']
        self.log(f"Scraping AllAfrica for {country}")
        
        articles = response.css('.container.mid .row .col-tn-12.col-sm-8.column.main .section.box.headlines.two-column .content .stories li a')
        
        for article in articles:
            title = article.css('::attr(title)').get()
            link = article.css('::attr(href)').get()
            
            if link and not link.startswith('http'):
                link = response.urljoin(link)
            
            yield scrapy.Request(url=link, callback=self.parse_article, meta={'title': title, 'country': country})
            time.sleep(random.uniform(0.5, 1))  # Random delay between requests

    def parse_article(self, response):
        title = response.meta['title']
        country = response.meta['country']
        
        paragraphs = response.css('.container.mid .row .col-tn-12.col-sm-8.column.main .story-body p::text').getall()
        
        body = ' '.join(paragraphs).strip()
        
        org_link = response.css('.container.mid .row .col-tn-12.col-sm-8.column.main .story-footer-link .source-url::attr(href)').get()

        self.data.append({
            'TITLE': title,
            'COUNTRY': country,
            'BODY': body,
            'Website Link': org_link
        })

    def closed(self, reason):
        df = pd.DataFrame(self.data)
        conn = sql.connect('newsData.db')
        df.to_csv('newsData.csv', index=False)
        df.to_sql('news', conn, if_exists='replace', index=False)
        conn.close()

**Create a .py file called runspider to run the spider**

In [3]:
from scrapy.crawler import CrawlerRunner
from scrapy.utils.project import get_project_settings
from twisted.internet import reactor
from twisted.internet.defer import inlineCallbacks
from newsspider.spiders.news_scraper import TrendsNewsSpider
from scrapy.utils.log import configure_logging
import sys

# Configure logging for Scrapy
configure_logging()

@inlineCallbacks
def crawl(runner, country):
    yield runner.crawl(TrendsNewsSpider, country=country)
    reactor.stop()

def run_spider(country):
    runner = CrawlerRunner(get_project_settings())
    crawl(runner, country)
    reactor.run(installSignalHandlers=False)

if __name__ == "__main__":
    country = sys.argv[1]
    run_spider(country)

### Natural Language Processing

# Phase 2           
Display News

The self-created rubric, in our repository, explains the requirement for a proper execution of this phase as seen below.   
**Description:** Show top 10 news headlines with one paragraph summaries and allow user to select and
read desired articles

### Repository Link 

Here is a link to our repository:

[What's The Hot Topic In Town?](https://github.com/kelvin-ahiakpor/Whats.The.Hot.Topic.In.Town)

### Imports

In [1]:
import re
import pandas as pd
import ipywidgets as widgets 

from collections import Counter
from transformers import pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS

**Load the default sentiment analysis pipeline**   
`distilbert-base-uncased-finetuned-sst-2-english`

In [2]:
sentiment_analyzer = pipeline("sentiment-analysis", 
                              model="distilbert/distilbert-base-uncased-finetuned-sst-2-english")

Hardware accelerator e.g. GPU is available in the environment, but no `device` argument is passed to the `Pipeline` object. Model will be on CPU.


**Widget creation**

In [3]:
upload_widget = widgets.FileUpload(multiple=False)
upload_button = widgets.Button(description="Process Upload")

### Upload a news data csv

Here we define a set of functions to allow uploads

**The following set of functions allow uploads in both Jupyter and Google Colab.**  
**Note:** This is for demonstration purposes only.  
In our deployed app this process is automated where the news data csv is saved to a sqllite database after scraping with scrapy and is loaded back into the application to perform natural language processing and identify the top 10 news articles.

**Global variables**  
Dictionary of upload information   
Video path  
Upload Widget
Upload Status

In [4]:
upload_info = None
file_path = None

In [5]:
upload_completed = False

In [6]:
def validate_file_upload(uploaded_files):
    for filename, file_info in uploaded_files.items():
        save_file(filename, file_info['content'])
        return filename

In [7]:
def save_file(filename, file_content):
    with open(filename, 'wb') as f:
        f.write(file_content)

In [8]:
def process_files(button):
    global upload_completed
    global file_path
    uploaded_files = upload_widget.value
    # Handle file validation and save
    file_path = validate_file_upload(uploaded_files)
    if file_path:
        print(f"File path set to: {file_path}")
        upload_completed = True  # Mark the upload as complete
    else:
        print("No valid file uploaded.")
    
    if upload_completed:
        print()
        print("Upload complete. You can now proceed to the run the next cells.")

In [9]:
def show_upload_widget():
    global upload_button
    upload_button = widgets.Button(description="Upload File")
    upload_button.on_click(process_files)
    
    # Display the widgets
    display(upload_widget)
    display(upload_button)
    
    print("Please upload a file and then click the 'Upload File' button.")

In [10]:
show_upload_widget()

FileUpload(value={}, description='Upload')

Button(description='Upload File', style=ButtonStyle())

Please upload a file and then click the 'Upload File' button.
File path set to: newsData.csv

Upload complete. You can now proceed to the run the next cells.


### Load the uploaded news data file

In [11]:
if upload_completed:
    newsdata = pd.read_csv(file_path)
    display(newsdata)
else:
    print("Upload a file first and then proceed to load the data.")

TITLE   COUNTRY  \
0   Zimbabwean in South Africa Challenges Exorbita...  zimbabwe   
1   American Star Tiffany Haddish Gives Zimbabwe a...  zimbabwe   
2   Insect Farming New Frontier for Smallholder Li...  zimbabwe   
3                    Zanu-PF, Swapo in Key Engagement  zimbabwe   
4                                                 NaN  zimbabwe   
5   Zanu-PF Official in Court Over $4 Million Frau...  zimbabwe   
6   Despite Hostility, LGBTQI+ Activists in Zimbab...  zimbabwe   
7         Zim Swimmer Weisthuizen Out of the Olympics  zimbabwe   
8                          Dynamos' Mangombe in a Fix  zimbabwe   
9             UN Tourism Delegation Visits First Lady  zimbabwe   
10            VP Chiwenga Rallies SADC to Boost Trade  zimbabwe   
11      VP Mohadi in Iran for Pezeshkian Inauguration  zimbabwe   
12                          Wheat Clusters Take Shape  zimbabwe   
13                              Tyla Captivates Crowd  zimbabwe   
14              Tourism Receipts Up 35pc to U.S.$241m  zimbabwe   
15              Air Zim Resumes Harare-Joburg Flights  zimbabwe   
16  Air Ambulances Training Starts. . .300 Special...  zimbabwe   
17  AfCFTA Pursues Exports for Africa's Food Security  zimbabwe   
18          The Essential Role of Liability Insurance  zimbabwe   
19                    Sables Lift Up Nation's Spirits  zimbabwe   
20       Afreximbank Earmarks U.S.$400m for Zim Firms  zimbabwe   
21  Presidential Solar Scheme Innovative, Imaginative  zimbabwe   
22  US Diplomat Runs Over 11-Year-Old Girl, Flees ...  zimbabwe   
23                         Police Hunt for 8 Suspects  zimbabwe   
24                                Adachi Finding Form  zimbabwe   
25  'Scrapping Vat On Cattle Sales Will Unlock Tru...  zimbabwe   
26           ZRP Bans Dangerous Weapons in Beitbridge  zimbabwe   
27      Macheso Names 'Kupa Kuturika' Album Tracklist  zimbabwe   
28         Pakare Paye Product Relishes Tuku Festival  zimbabwe   
29              78 CCC Activists Want Charges Quashed  zimbabwe   
30  Mnangagwa Taking Over SADC Chairmanship a Trav...  zimbabwe   
31  Police Crackdown On Vendors, Touts, Car Wash O...  zimbabwe   
32  Mutapa Fund Injects U.S.$5 Million Into Invict...  zimbabwe   

                                                 BODY  \
0   A Zimbabwean resident in South Africa, Takudzw...   
1   VISITING American comedienne, Tiffany Haddish,...   
2   Zimbabwe, just like many other countries, is l...   
3   ZANU PF and SWAPO of Namibia share common bond...   
4                                                 NaN   
5   A senior official from Zimbabwe's ruling Zanu ...   
6   Munich, Germany —  LGBTQI+ people have long be...   
7   ZIMBABWE swimmer Paige Van Der Weisthuizen was...   
8   Castle Lager Premiership giants, Dynamos head ...   
9   Victoria Falls — United Nations Tourism Secret...   
10  SADC countries must use their regional advanta...   
11  Tehran, Iran — Vice President Kembo Mohadi has...   
12  CLUSTERING of wheat farmers for easy service d...   
13  Johannesburg. — Tyla captivated the Paris crow...   
14  Zimbabwe's tourism receipts grew by an estimat...   
15  Air Zimbabwe yesterday resumed flights from Ha...   
16  THE Russian HeliDrive Air ambulances medical s...   
17  THE African Continental Free Trade Area (AfCFT...   
18  Liability insurance is a critical safeguard fo...   
19  After breaking a couple of jinxes to be crowne...   
20  The African Export-Import Bank (Afreximbank) h...   
21  This is an installation by Zimbabwe Solar Ener...   
22  A United States Embassy official allegedly ran...   
23  POLICE in Harare have launched a manhunt for e...   
24  Adachi, who had a poor start to the Zifa South...   
25  Stakeholders in the beef industry have describ...   
26  The Zimbabwe Republic Police (ZRP) has issued ...   
27  ALICK Macheso's management has finally release...   
28  SEASONED arts consultant and publicist, Eugine...   
29  Lawyers representing the 78 Citizens Coalition...   
30

In [12]:
newsdata

TITLE   COUNTRY  \
0   Zimbabwean in South Africa Challenges Exorbita...  zimbabwe   
1   American Star Tiffany Haddish Gives Zimbabwe a...  zimbabwe   
2   Insect Farming New Frontier for Smallholder Li...  zimbabwe   
3                    Zanu-PF, Swapo in Key Engagement  zimbabwe   
4                                                 NaN  zimbabwe   
5   Zanu-PF Official in Court Over $4 Million Frau...  zimbabwe   
6   Despite Hostility, LGBTQI+ Activists in Zimbab...  zimbabwe   
7         Zim Swimmer Weisthuizen Out of the Olympics  zimbabwe   
8                          Dynamos' Mangombe in a Fix  zimbabwe   
9             UN Tourism Delegation Visits First Lady  zimbabwe   
10            VP Chiwenga Rallies SADC to Boost Trade  zimbabwe   
11      VP Mohadi in Iran for Pezeshkian Inauguration  zimbabwe   
12                          Wheat Clusters Take Shape  zimbabwe   
13                              Tyla Captivates Crowd  zimbabwe   
14              Tourism Receipts Up 35pc to U.S.$241m  zimbabwe   
15              Air Zim Resumes Harare-Joburg Flights  zimbabwe   
16  Air Ambulances Training Starts. . .300 Special...  zimbabwe   
17  AfCFTA Pursues Exports for Africa's Food Security  zimbabwe   
18          The Essential Role of Liability Insurance  zimbabwe   
19                    Sables Lift Up Nation's Spirits  zimbabwe   
20       Afreximbank Earmarks U.S.$400m for Zim Firms  zimbabwe   
21  Presidential Solar Scheme Innovative, Imaginative  zimbabwe   
22  US Diplomat Runs Over 11-Year-Old Girl, Flees ...  zimbabwe   
23                         Police Hunt for 8 Suspects  zimbabwe   
24                                Adachi Finding Form  zimbabwe   
25  'Scrapping Vat On Cattle Sales Will Unlock Tru...  zimbabwe   
26           ZRP Bans Dangerous Weapons in Beitbridge  zimbabwe   
27      Macheso Names 'Kupa Kuturika' Album Tracklist  zimbabwe   
28         Pakare Paye Product Relishes Tuku Festival  zimbabwe   
29              78 CCC Activists Want Charges Quashed  zimbabwe   
30  Mnangagwa Taking Over SADC Chairmanship a Trav...  zimbabwe   
31  Police Crackdown On Vendors, Touts, Car Wash O...  zimbabwe   
32  Mutapa Fund Injects U.S.$5 Million Into Invict...  zimbabwe   

                                                 BODY  \
0   A Zimbabwean resident in South Africa, Takudzw...   
1   VISITING American comedienne, Tiffany Haddish,...   
2   Zimbabwe, just like many other countries, is l...   
3   ZANU PF and SWAPO of Namibia share common bond...   
4                                                 NaN   
5   A senior official from Zimbabwe's ruling Zanu ...   
6   Munich, Germany —  LGBTQI+ people have long be...   
7   ZIMBABWE swimmer Paige Van Der Weisthuizen was...   
8   Castle Lager Premiership giants, Dynamos head ...   
9   Victoria Falls — United Nations Tourism Secret...   
10  SADC countries must use their regional advanta...   
11  Tehran, Iran — Vice President Kembo Mohadi has...   
12  CLUSTERING of wheat farmers for easy service d...   
13  Johannesburg. — Tyla captivated the Paris crow...   
14  Zimbabwe's tourism receipts grew by an estimat...   
15  Air Zimbabwe yesterday resumed flights from Ha...   
16  THE Russian HeliDrive Air ambulances medical s...   
17  THE African Continental Free Trade Area (AfCFT...   
18  Liability insurance is a critical safeguard fo...   
19  After breaking a couple of jinxes to be crowne...   
20  The African Export-Import Bank (Afreximbank) h...   
21  This is an installation by Zimbabwe Solar Ener...   
22  A United States Embassy official allegedly ran...   
23  POLICE in Harare have launched a manhunt for e...   
24  Adachi, who had a poor start to the Zifa South...   
25  Stakeholders in the beef industry have describ...   
26  The Zimbabwe Republic Police (ZRP) has issued ...   
27  ALICK Macheso's management has finally release...   
28  SEASONED arts consultant and publicist, Eugine...   
29  Lawyers representing the 78 Citizens Coalition...   
30

In [13]:
newsdata = newsdata.dropna()
newsdata_body = newsdata['BODY']

**Function to split text into sentences**

In [14]:
def split_text_by_sentences(text):
    sentences = re.split(r'(?<!\w\.\w.)(?<![A-Z][a-z]\.)(?<=\.|\?)\s', text)
    return sentences

**Preprocess the data by converting it to lower case** 

In [15]:
def preprocess(text):
    text = text.lower()  # Lowercase
    # Add more preprocessing steps if needed
    return text

In [16]:
newsdata.loc[:, 'cleaned_body'] = newsdata['BODY'].apply(preprocess)

/var/folders/1d/qzlhmy690tvcd1wf5tpyr_km0000gn/T/ipykernel_13991/3866009567.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  newsdata.loc[:, 'cleaned_body'] = newsdata['BODY'].apply(preprocess)


### Use TF-IDF to retrieve most important terms in the news data and select top 10 news articles
The TFidVectorizer is sued to transform the body of the articles into numeric form, finding the top term based on their term frequencies. These terms aare then counted in each article, where the articles are arranged based on highest number of terms present.

**Vectorize the text**

In [17]:
vectorizer = TfidfVectorizer(stop_words=list(ENGLISH_STOP_WORDS), max_features=1000)
X = vectorizer.fit_transform(newsdata['cleaned_body'])

**Calculate term frequencies**

In [18]:
feature_names = vectorizer.get_feature_names_out()
tfidf_scores = X.sum(axis=0).A1
term_scores = dict(zip(feature_names, tfidf_scores))

**Get top 10 trending topics based on TF-IDF scores using the 20 most common terms**

In [19]:
top_terms = Counter(term_scores).most_common(20)

In [20]:
print("Top Relevant Terms:")
for term, score in top_terms:
    print(f"{term}: {score}")

article_scores = []

# Calculate relevance score for each article based on top terms
for index, row in newsdata.iterrows():
    relevance_score = 0
    for term, _ in top_terms:
        relevance_score += row['cleaned_body'].count(term)  # Count occurrences of the term
    article_scores.append({'TITLE': row['TITLE'],
                           'URL': row['Website Link'],
                           'BODY':row['BODY'],
                           'Relevance': relevance_score,
                           })

Top Relevant Terms:
said: 1.617239418765897
sadc: 1.5866304053154328
zimbabwe: 1.5387162220256676
africa: 1.0980387025610683
investment: 1.0527951933568427
tourism: 1.0083783864340194
harare: 0.9963720158183158
country: 0.9404634824655383
police: 0.9276228224586293
million: 0.9266161434507789
region: 0.9027383259263309
season: 0.8665253037480395
mr: 0.8141305710770922
president: 0.7944616990579325
government: 0.742501399340561
year: 0.7418823191337467
percent: 0.7307642736221417
international: 0.7276504583290118
mnangagwa: 0.7108881419472728
economic: 0.707578804741229


In [21]:
# Convert list to DataFrame
article_scores_df = pd.DataFrame(article_scores)

# Sort articles by relevance score and get top 10
top_10_articles = article_scores_df.sort_values(by='Relevance', ascending=False).head(10)

In [22]:
top_10_articles

TITLE  \
18       Afreximbank Earmarks U.S.$400m for Zim Firms   
8             VP Chiwenga Rallies SADC to Boost Trade   
28  Mnangagwa Taking Over SADC Chairmanship a Trav...   
12              Tourism Receipts Up 35pc to U.S.$241m   
20  US Diplomat Runs Over 11-Year-Old Girl, Flees ...   
15  AfCFTA Pursues Exports for Africa's Food Security   
7             UN Tourism Delegation Visits First Lady   
14  Air Ambulances Training Starts. . .300 Special...   
13              Air Zim Resumes Harare-Joburg Flights   
29  Police Crackdown On Vendors, Touts, Car Wash O...   

                                                  URL  \
18  https://www.herald.co.zw/afreximbank-earmarks-...   
8   https://www.herald.co.zw/vp-chiwenga-rallies-s...   
28  https://www.newzimbabwe.com/mnangagwa-taking-o...   
12  https://www.herald.co.zw/tourism-receipts-up-3...   
20  https://www.herald.co.zw/us-diplomat-runs-over...   
15  https://www.herald.co.zw/afcfta-pursues-export...   
7   https://www.herald.co.zw/un-tourism-delegation...   
14  https://www.herald.co.zw/air-ambulances-traini...   
13  https://www.herald.co.zw/air-zim-resumes-harar...   
29  https://www.newzimbabwe.com/police-crackdown-o...   

                                                 BODY  Relevance  
18  The African Export-Import Bank (Afreximbank) h...        136  
8   SADC countries must use their regional advanta...         77  
28  ZIMBABWEANS living abroad have raised concern ...         75  
12  Zimbabwe's tourism receipts grew by an estimat...         63  
20  A United States Embassy official allegedly ran...         52  
15  THE African Continental Free Trade Area (AfCFT...         49  
7   Victoria Falls — United Nations Tourism Secret...         48  
14  THE Russian HeliDrive Air ambulances medical s...         44  
13  Air Zimbabwe yesterday resumed flights from Ha...         32  
29  The Zimbabwe Republic Police (ZRP) in conjunct...         28

### Natural Language Processing

# Phase 3           
Sentiment Analysis

**Description:** Apply the distilBERT base uncased SST-2 sentiment analysis tool from using Hugging Face, to
understand and display the emotional tone of news articles, either negative or positive.   

#### About distilBERT
The distilbert-base-uncased-finetuned-sst-2-english is applied on the dataset(more specifically the body of the articles to acquire the sentiment analysis of the article in question. This is done by splitting the articles into chunks and passing this chunks into the model, the sentiment scores of all these are articles are averaged and the average score of either negative or positive is compared, The highest average is then selected, then displaying the sentiment analysis as either Positive or negative. "

In [23]:
for text in top_10_articles['BODY']:
    sentences = split_text_by_sentences(text)
    results = []

    # Analyze each sentence and calculate the average sentiment score
    for sentence in sentences:
        result = sentiment_analyzer(sentence)[0]
        results.append(result)
        #print(f"Sentence: {sentence}\nSentiment: {result['label']}, Probability: {result['score']:.4f}\n")
 
    # Calculate the average sentiment score
    positive_scores = [result['score'] for result in results if result['label'] == 'POSITIVE']
    negative_scores = [result['score'] for result in results if result['label'] == 'NEGATIVE']

    average_positive = sum(positive_scores) / len(positive_scores) if positive_scores else 0
    average_negative = sum(negative_scores) / len(negative_scores) if negative_scores else 0

    # Determine the overall average sentiment
    overall_sentiment = "POSITIVE" if average_positive >= average_negative else "NEGATIVE"
    average_score = max(average_positive, average_negative)

    for index, row in top_10_articles.iterrows():
        if text in row['BODY']:
            print()
            print("TITLE OF ARTICLE: "+ row['TITLE'])

    print(f"Overall Sentiment: {overall_sentiment}")
    print(f"Average Sentiment Score: {average_score:.4f}")

    '''# Display the probabilities for each sentence
    for i, result in enumerate(results):
      print(f"Sentence {i+1} Sentiment: {result['label']}, Probability: {result['score']:.4f}")'''


TITLE OF ARTICLE: Afreximbank Earmarks U.S.$400m for Zim Firms
Overall Sentiment: POSITIVE
Average Sentiment Score: 0.9633

TITLE OF ARTICLE: VP Chiwenga Rallies SADC to Boost Trade
Overall Sentiment: POSITIVE
Average Sentiment Score: 0.9920

TITLE OF ARTICLE: Mnangagwa Taking Over SADC Chairmanship a Travesty - Diasporans Accuse Regional Bloc of Legitimizing Tyranny
Overall Sentiment: NEGATIVE
Average Sentiment Score: 0.9585

TITLE OF ARTICLE: Tourism Receipts Up 35pc to U.S.$241m
Overall Sentiment: POSITIVE
Average Sentiment Score: 0.9916

TITLE OF ARTICLE: US Diplomat Runs Over 11-Year-Old Girl, Flees Country
Overall Sentiment: NEGATIVE
Average Sentiment Score: 0.9706

TITLE OF ARTICLE: AfCFTA Pursues Exports for Africa's Food Security
Overall Sentiment: POSITIVE
Average Sentiment Score: 0.9948

TITLE OF ARTICLE: UN Tourism Delegation Visits First Lady
Overall Sentiment: NEGATIVE
Average Sentiment Score: 0.9741

TITLE OF ARTICLE: Air Ambulances Training Starts. . .300 Specialised J

### Natural Language Processing

# Phase 4         
Summarize News Articles

This notebook addresses Phase 4 of the What's The Hot Topic in Topic Town? project: **Summarize News Articles**.  
The self-created rubric, in our repository, explains the requirement for a proper execution of this phase as seen below.   
**Description:** Fine tune the BERT base cased model from Hugging Face with labeled dataset of news articles
and generate one paragraph news summaries with it

### Repository Link

Here is a link to our repository:

[What's The Hot Topic In Town?](https://github.com/kelvin-ahiakpor/Whats.The.Hot.Topic.In.Town)

### Imports

In [ ]:
%%capture
!pip install datasets==1.0.2
!pip install transformers==4.2.1
!pip install torch evaluate accelerate pandas rouge_score wandb sacrebleu wandb

import os
import wandb
import torch
import shutil
import evaluate
import datasets
import accelerate
import transformers

import pandas as pd
from torch.cuda.amp import autocast
from IPython.display import display, HTML
from datasets import ClassLabel
from datasets import load_metric

from transformers import pipeline
from transformers import BertTokenizerFast
from transformers import EncoderDecoderModel
from transformers import Seq2SeqTrainingArguments
from transformers import Seq2SeqTrainer

from IPython.display import clear_output

from dataclasses import dataclass, field
from typing import Optional

### Setting job timeout for computation

In [ ]:
os.environ['JOBLIB_START_METHOD'] = 'loky'
os.environ['JOBLIB_TIMEOUT'] = '300'

### Setting up weights and biases (WANDB) to log training metrics

In [ ]:
os.environ['WANDB_API_KEY'] = 'removed-for-privacy'
wandb.login()

wandb: Currently logged in as: kelvin-ahiakpor (kelvin-ahiakpor-ashesi-university). Use `wandb login --relogin` to force relogin


True

### Selecting first available GPU if available

In [ ]:
device = 0 if torch.cuda.is_available() else -1

### Loading training data

In [ ]:
train_data = datasets.load_dataset("cnn_dailymail", "3.0.0", split="train")

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Generating train split:   0%|          | 0/287113 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/13368 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/11490 [00:00<?, ? examples/s]

### Custom recipes for DataFrame inspection
Object predictions will be stored in a dataframe so it can be displayed in a friendly manner in this notebook

In [ ]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)
pd.set_option('display.precision', 3)

### Understanding the data

##### Inspecting dataset's metadata

In [ ]:
train_data

Dataset({
    features: ['article', 'highlights', 'id'],
    num_rows: 287113
})

In [ ]:
train_data.features

{'article': Value(dtype='string', id=None),
 'highlights': Value(dtype='string', id=None),
 'id': Value(dtype='string', id=None)}

In [ ]:
train_data.info.splits

{'train': SplitInfo(name='train', num_bytes=1261703785, num_examples=287113, shard_lengths=[115705, 115704, 55704], dataset_name='cnn_dailymail'),
 'validation': SplitInfo(name='validation', num_bytes=57732412, num_examples=13368, shard_lengths=None, dataset_name='cnn_dailymail'),
 'test': SplitInfo(name='test', num_bytes=49925732, num_examples=11490, shard_lengths=None, dataset_name='cnn_dailymail')}

So far we see our dataset is already split into train, validation and test with 287113, 13368, 11490 examples respectively

##### Peek at training data

Our input is called article and our labels are called highlights. Let's now print out the first example of the training data to get a feeling for the data.

In [ ]:
df = pd.DataFrame(train_data[:1])
del df["id"]
for column, typ in train_data.features.items():
      if isinstance(typ, ClassLabel):
            df[column] = df[column].transform(lambda i: typ.names[i])
display(HTML(df.to_html()))

The input data seems to consist of short news articles. Interestingly, the labels appear to be bullet-point-like summaries. At this point, we should probably take a look at a couple of other examples to get a better feeling for the data.

In [ ]:
df = pd.DataFrame(train_data[1:4])
del df["id"]
for column, typ in train_data.features.items():
      if isinstance(typ, ClassLabel):
            df[column] = df[column].transform(lambda i: typ.names[i])
display(HTML(df.to_html()))

Next, let's get a sense of the length of input data and labels.  
As models compute length in *token-length*, we will make use of the `bert-base-cased` tokenizer to compute the article and summary length.

##### Quick statistics
Artilce and Summary Length,
Histogram Plots,
Correlation Matrix,
Categorical Variable Selection &
Important Numeric Features from PCA

In [ ]:
tokenizer = BertTokenizerFast.from_pretrained("bert-base-cased")

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

Next, we make use of `.map()` to compute the length of the article and its summary. The maximum length that `bert-base-uncased` can process amounts to 512, we are also interested in the percentage of input samples being longer than the maximum length.
Similarly, we compute the percentage of summaries that are longer than 64, and 128 respectively as these will help us meet our goal of providing multi-paragraph summaries if possible

We can define the `.map()` function as follows.

In [ ]:
# map article and summary len to dict as well as if sample is longer than 512 tokens
def map_to_length(x):
    x["article_len"] = len(tokenizer(x["article"]).input_ids)
    x["article_longer_512"] = int(x["article_len"] > tokenizer.model_max_length)
    x["summary_len"] = len(tokenizer(x["highlights"]).input_ids)
    x["summary_longer_64"] = int(x["summary_len"] > 64)
    x["summary_longer_128"] = int(x["summary_len"] > 128)
    return x

It should be sufficient to look at the first 10000 samples. We can speed up the mapping by using multiple processes with `num_proc=4`.

In [ ]:
sample_size = 10000
data_stats = train_data.select(range(sample_size)).map(map_to_length, num_proc=4)

/usr/local/lib/python3.10/dist-packages/multiprocess/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  self.pid = os.fork()


Map (num_proc=4):   0%|          | 0/10000 [00:00<?, ? examples/s]

Having computed the length for the first 10000 samples, we can now average them together. We can make use of the `.map()` function with `batched=True` and `batch_size=-1` to have access to all 10000 samples within the `.map()` function.

We have also identified that some articles are too long and can cause training errors we will `use.filter()` to remove them in a data cleaning section below

In [ ]:
def compute_and_print_stats(x):
    if len(x["article_len"]) == sample_size:
        print(
            "Article Mean: {}, %-Articles > 512:{}, Summary Mean:{}, %-Summary > 64:{}, %-Summary > 128:{}".format(
                sum(x["article_len"]) / sample_size,
                sum(x["article_longer_512"]) / sample_size,
                sum(x["summary_len"]) / sample_size,
                sum(x["summary_longer_64"]) / sample_size,
                sum(x["summary_longer_128"]) / sample_size,
            )
        )

output = data_stats.map(compute_and_print_stats, batched=True, batch_size=-1,)

Map:   0%|          | 0/10000 [00:00<?, ? examples/s]

Article Mean: 827.079, %-Articles > 512:0.7344, Summary Mean:59.5725, %-Summary > 64:0.3499, %-Summary > 128:0.0


We can see that on average an article contains 827 tokens with *ca.* 3/4 of the articles being longer than the model's `max_length` 512. The summary is on average 59 tokens long. About 35% of our 10000-sample summaries are longer than 64 tokens, but none are longer than 128 tokens.

`bert-base-cased` is limited to 512 tokens, which means we would have to cut possibly important information from the article. Because most of the important information is often found at the beginning of articles and because we want to be computationally efficient, we decide to stick to `bert-base-cased` with a `max_length` of 512 in this notebook. This choice is not optimal but has shown to yield [good results](https://arxiv.org/abs/1907.12461) on CNN/Dailymail.


#### Some notes so far

1. Evaluation :   

    **Case-sensitivity**
    - The text in our peeks were *case-sensitive*. This means that we have to be careful if we want to use *case-insensitive* models. As *CNN/Dailymail* is a summarization dataset, the model will be evaluated using the *ROUGE* metric. Checking the description of *ROUGE* in 🤗datasets, *cf.* [here](https://huggingface.co/metrics/rouge), we can see that the metric is *case-insensitive*, meaning that *upper case* letters will be normalized to *lower case* letters during evaluation. Thus, we can safely leverage *uncased* checkpoints, such as `bert-base-uncased`.

# Task 1
Test the pretrained BERT-Based model for summarization

Why use bert2BERT and not BERT_BASE?  
bert2BERT is a Reusable Pretrained Language Models
In recent years, researchers tend to pre-train ever-larger language models to explore the upper limit of deep models. However, large language model pre-training costs intensive computational resources and most of the models are trained from scratch without reusing the existing pre-trained models, which is wasteful. bert2BERT allows us to  effectively transfer the knowledge of an existing smaller pre-trained model like BERT_BASE to a large model  through parameter initialization and significantly improve the pre-training efficiency of the large model. [1]

**Loading a current default summarization model bart-large from the summarization pipeline**

In [ ]:
summarizer = pipeline("summarization", model="facebook/bart-large-cnn",device=device)

config.json:   0%|          | 0.00/1.58k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

In [ ]:
def summarize_article(article, tokenizer, summarizer):
    inputs = tokenizer(article, return_tensors="pt", truncation=True, max_length=512)
    inputs = {key: val.to("cuda") for key, val in inputs.items()}  # Ensure inputs are on GPU
    summary = summarizer(article, max_length=150, min_length=50, do_sample=True)
    return summary[0]['summary_text']

**Generating a few sample summaries with it**

In [ ]:
news_articles = [example["article"] for example in train_data.select(range(3))]
target_summaries = [example["highlights"] for example in train_data.select(range(3))]

Summarize the articles and store summaries in summaries list. We set `do_sample` to `True` to avoid greedy decoding and provide a more natural, human-like summary.

In [ ]:
pre_train_summaries = []
for article in news_articles:
    try:
        summary = summarize_article(article, tokenizer, summarizer)
        pre_train_summaries.append(summary)
    except Exception as e:
        pre_train_summaries.append("Error, article was too long")

Create a dataframe containing the pretrained BART summaries and the target summaries from our fine-tuning dataset

In [ ]:
summaries = pd.DataFrame({
    'Pre-trained BART Summaries': pre_train_summaries,
    'Target Summaries': target_summaries
})

In [ ]:
summaries

Pre-trained BART Summaries  \
0                                                                       Harry Potter star Daniel Radcliffe turns 18 on Monday. He gains access to a reported £20 million ($41.1 million) fortune. Radcliffe's earnings from the first five Potter films have been held in a trust fund. Details of how he'll mark his landmark birthday are under wraps.   
1                                                            Judge Steven Leifman is an advocate for justice and the mentally ill. About one-third of all people in Miami-Dade county jails are mentally ill, he says. He says the sheer volume is overwhelming the system. Starting in 2008, many inmates will be sent to a new mental health facility.   
2  NEW: "I probably had a 30-, 35-foot free fall," survivor Gary Babineau says. NEW: "My truck was completely face down, pointed toward the ground, and my truck got ripped in half," he says. Dr. John Hink says he saw dozens of people lying dazed on an expansive deck. "It just gave a rumble real quick, and it all just gave way," survivor says.   

                                                                                                                                                                                                                                                                               Target Summaries  
0                                                                   Harry Potter star Daniel Radcliffe gets £20M fortune as he turns 18 Monday .\nYoung actor says he has no plans to fritter his cash away .\nRadcliffe's earnings from first five Potter films have been held in trust fund .  
1  Mentally ill inmates in Miami are housed on the "forgotten floor"\nJudge Steven Leifman says most are there as a result of "avoidable felonies"\nWhile CNN tours facility, patient shouts: "I am the son of the president"\nLeifman says the system is unjust and he's fighting for change .  
2                                                           NEW: "I thought I was going to die," driver says .\nMan says pickup truck was folded in half; he just has cut on face .\nDriver: "I probably had a 30-, 35-foot free fall"\nMinnesota bridge collapsed during rush hour Wednesday .

These are good summaries and this is what we will be finetuning bert-base-uncased to do in the next task.

# Task 2
Fine-Tune bert-base-uncased as bert2BERTMK on the cnn_dailymail dataset
KM - Kelvin and Manuel

**Note**: We are using the cnn daily mail dataset because it is tailored for abstractive summarization where as the BBC Xsum dataset is for extreme summarization. We need abstractive because it can help us achieve our goal of multi-sentence summaries.

### Loading validation data

First, we load 10% of the validation dataset for faster validation:

In [ ]:
val_data = datasets.load_dataset("cnn_dailymail", "3.0.0", split="validation[:10%]")

### Data Preprocessing

In [ ]:
tokenizer.bos_token = tokenizer.cls_token
tokenizer.eos_token = tokenizer.sep_token

In [ ]:
batch_size=4
encoder_max_length=512
decoder_max_length=128

In [ ]:
def process_data_to_model_inputs(batch):
  # tokenize the inputs and labels
    inputs = tokenizer(batch["article"], padding="max_length", truncation=True, max_length=encoder_max_length)
    outputs = tokenizer(batch["highlights"], padding="max_length", truncation=True, max_length=decoder_max_length)

    batch["input_ids"] = inputs.input_ids
    batch["attention_mask"] = inputs.attention_mask
    batch["labels"] = outputs.input_ids.copy()

    # because BERT automatically shifts the labels, the labels correspond exactly to `decoder_input_ids`.
    # We have to make sure that the PAD token is ignored
    batch["labels"] = [[-100 if token == tokenizer.pad_token_id else token for token in labels] for labels in
                       batch["labels"]]

    return batch

In [ ]:
# only use 32 training examples for notebook - DELETE LINE FOR FULL TRAINING
train_data = train_data.select(range(140000)) #70000

train_data = train_data.map(
    process_data_to_model_inputs,
    batched=True,
    batch_size=batch_size,
    remove_columns=["article", "highlights", "id"]
)
train_data.set_format(
    type="torch", columns=["input_ids", "attention_mask", "labels"],
)

Map:   0%|          | 0/140000 [00:00<?, ? examples/s]

In [ ]:
# only use 16 training examples for notebook - DELETE LINE FOR FULL TRAINING
#val_data = val_data.select(range(1200))

val_data = val_data.map(
    process_data_to_model_inputs,
    batched=True,
    batch_size=batch_size,
    remove_columns=["article", "highlights", "id"]
)
val_data.set_format(
    type="torch", columns=["input_ids", "attention_mask", "labels"],
)

Map:   0%|          | 0/1337 [00:00<?, ? examples/s]

### Warm-starting the Encoder-Decoder Model
The error messages are normal. they show that some weights are being randomly initialized

In [ ]:
%%capture
bert2bert = EncoderDecoderModel.from_encoder_decoder_pretrained("bert-base-cased", "bert-base-cased")

In [ ]:
# set special tokens
bert2bert.config.decoder_start_token_id = tokenizer.bos_token_id
bert2bert.config.eos_token_id = tokenizer.eos_token_id
bert2bert.config.pad_token_id = tokenizer.pad_token_id

# sensible parameters for beam search
bert2bert.config.vocab_size = bert2bert.config.decoder.vocab_size
bert2bert.config.max_length = 142
bert2bert.config.min_length = 56
bert2bert.config.no_repeat_ngram_size = 3
bert2bert.config.early_stopping = True
bert2bert.config.length_penalty = 2.0
bert2bert.config.num_beams = 4

### **Fine-Tuning Warm-Started Encoder-Decoder Models**

For the `EncoderDecoderModel` framework, we will use the `Seq2SeqTrainingArguments` and the `Seq2SeqTrainer` which have been imported already.

### Load rouge for validation

In [ ]:
rouge = datasets.load_metric("rouge", trust_remote_code=True)

<ipython-input-32-c4923d78972b>:1: FutureWarning: load_metric is deprecated and will be removed in the next major version of datasets. Use 'evaluate.load' instead, from the new library 🤗 Evaluate: https://huggingface.co/docs/evaluate
  rouge = datasets.load_metric("rouge", trust_remote_code=True)


In [ ]:
def compute_metrics(pred):
    labels_ids = pred.label_ids
    pred_ids = pred.predictions

    # all unnecessary tokens are removed
    pred_str = tokenizer.batch_decode(pred_ids, skip_special_tokens=True)
    labels_ids[labels_ids == -100] = tokenizer.pad_token_id
    label_str = tokenizer.batch_decode(labels_ids, skip_special_tokens=True)

    rouge_output = rouge.compute(predictions=pred_str, references=label_str, rouge_types=["rouge2"])["rouge2"].mid

    return {
        "rouge2_precision": round(rouge_output.precision, 4),
        "rouge2_recall": round(rouge_output.recall, 4),
        "rouge2_fmeasure": round(rouge_output.fmeasure, 4),
    }

**Setting training arguments**

For our model training, we referred to Patrick von Platen's demo of the BERT2BERT model to guide our parameter choices. The parameters initially suggested were as follows:

training_args = Seq2SeqTrainingArguments(  
    output_dir="./",  
    evaluation_strategy="steps",  
    per_device_train_batch_size=batch_size,  
    per_device_eval_batch_size=batch_size,  
    predict_with_generate=True,  
    logging_steps=2,  # set to 1000 for full training  
    save_steps=16,  # set to 500 for full training  
    eval_steps=4,  # set to 8000 for full training  
    warmup_steps=1,  # set to 2000 for full training  
    max_steps=16, # delete for full training  
    overwrite_output_dir=True,  
    save_total_limit=3,  
    fp16=True,  
)  

Due to GPU space limitations, we were unable to perform a grid search for hyperparameter tuning. Therefore, we had to adjust some of these parameters to fit our computational resources and training constraints. We consulted with ChatGPT and Faculty Intern Kweku Yamoah to fine-tune these settings.

In [ ]:
training_args = Seq2SeqTrainingArguments(
    output_dir="./",
    per_device_train_batch_size=batch_size,
    per_device_eval_batch_size=batch_size,
    predict_with_generate=True,
    evaluation_strategy = 'steps',
    do_train=True,
    do_eval=True,
    logging_steps=1000,  # set to 1000 for full training
    save_steps=1000,  # set to 500 for full training
    eval_steps=8000,  # set to 8000 for full training
    warmup_steps=2000,  # set to 2000 for full training
    overwrite_output_dir=True,
    save_total_limit=3,
    fp16=True,
    report_to="wandb"  # Enable W&B logging
)

/usr/local/lib/python3.10/dist-packages/transformers/training_args.py:1494: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(


**Instantiate trainer**

In [ ]:
device = "cuda:0" if torch.cuda.is_available() else "cpu"
bert2bert.to(device)

EncoderDecoderModel(
  (encoder): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(28996, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSdpaSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, el

In [ ]:
trainer = Seq2SeqTrainer(
    model=bert2bert,
    args=training_args,
    compute_metrics=compute_metrics,
    train_dataset=train_data,
    eval_dataset=val_data,
)

Cool! Finally, we start training.

**Train!**

In [ ]:
trainer.train()

/usr/local/lib/python3.10/dist-packages/transformers/models/encoder_decoder/modeling_encoder_decoder.py:622: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than tensor.new_tensor(sourceTensor).
  decoder_attention_mask = decoder_input_ids.new_tensor(decoder_input_ids != self.config.pad_token_id)
/usr/local/lib/python3.10/dist-packages/transformers/models/encoder_decoder/modeling_encoder_decoder.py:642: FutureWarning: Version v4.12.0 introduces a better way to train encoder-decoder models by computing the loss inside the encoder-decoder framework rather than in the decoder itself. You may observe training discrepancies if fine-tuning a model trained with versions anterior to 4.12.0. The decoder_input_ids are now created based on the labels, no need to pass them yourself anymore.
  warnings.warn(DEPRECATION_WARNING, FutureWarning)


/usr/local/lib/python3.10/dist-packages/transformers/models/encoder_decoder/modeling_encoder_decoder.py:622: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than tensor.new_tensor(sourceTensor).
  decoder_attention_mask = decoder_input_ids.new_tensor(decoder_input_ids != self.config.pad_token_id)
/usr/local/lib/python3.10/dist-packages/transformers/models/encoder_decoder/modeling_encoder_decoder.py:642: FutureWarning: Version v4.12.0 introduces a better way to train encoder-decoder models by computing the loss inside the encoder-decoder framework rather than in the decoder itself. You may observe training discrepancies if fine-tuning a model trained with versions anterior to 4.12.0. The decoder_input_ids are now created based on the labels, no need to pass them yourself anymore.
  warnings.warn(DEPRECATION_WARNING, FutureWarning)
/usr/local/lib/python3.10/dist-packages/tra

/usr/local/lib/python3.10/dist-packages/transformers/models/encoder_decoder/modeling_encoder_decoder.py:622: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than tensor.new_tensor(sourceTensor).
  decoder_attention_mask = decoder_input_ids.new_tensor(decoder_input_ids != self.config.pad_token_id)
/usr/local/lib/python3.10/dist-packages/transformers/models/encoder_decoder/modeling_encoder_decoder.py:642: FutureWarning: Version v4.12.0 introduces a better way to train encoder-decoder models by computing the loss inside the encoder-decoder framework rather than in the decoder itself. You may observe training discrepancies if fine-tuning a model trained with versions anterior to 4.12.0. The decoder_input_ids are now created based on the labels, no need to pass them yourself anymore.
  warnings.warn(DEPRECATION_WARNING, FutureWarning)
/usr/local/lib/python3.10/dist-packages/tra

TrainOutput(global_step=105000, training_loss=2.4967903680710566, metrics={'train_runtime': 24469.1844, 'train_samples_per_second': 17.164, 'train_steps_per_second': 4.291, 'total_flos': 2.5764875329536e+17, 'train_loss': 2.4967903680710566, 'epoch': 3.0})

The model achieves a ROUGE-2 score of **18.22**, which is even a little better than reported in the paper arvixx.  
**Be sure to edit this part**

# Task 3
Evaluate the fine-tuned model with the ROUGE-2 Metric

### Evaluation
**Rouge?** Rouge is an N-gram Co-Occurrence Statistic  
**Brief on Rouge**:
ROUGE-2 specifically measures the overlap of bigrams (two-word sequences) between the candidate summary and the reference summary.
Key Points of ROUGE-2
Bigram Overlap: ROUGE-2 evaluates the similarity based on two-word sequences present in both the candidate and reference summaries.

**Interpretation of Scores**
Higher ROUGE-2 Score:
Indicates better quality of the generated summary.
Suggests a higher degree of similarity between the bigrams in the generated summary and the reference summary.
Implies that the model is capturing more of the important bi-word sequences from the reference summary [4].  

We finished training our model. Let's now evaluate the model on the test data. We make use of the dataset's handy `.map()` function to generate a summary of each sample of the test data.

### Loading test data

In [ ]:
test_data = datasets.load_dataset("cnn_dailymail", "3.0.0", split="test")

In [ ]:
# only use 16 training examples for notebook - DELETE LINE FOR FULL TRAINING
test_data = test_data.select(range(5700))

### Load the fine tuned model

In [ ]:
tokenizer = BertTokenizerFast.from_pretrained("bert-base-cased")
bert2bertMK = EncoderDecoderModel.from_pretrained("/content/checkpoint-105000")
bert2bertMK.gradient_checkpointing_enable()
bert2bertMK.to("cuda")

EncoderDecoderModel(
  (encoder): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(28996, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, elemen

In [ ]:
batch_size = 16 #originally 16

# map data correctly
def generate_summary(batch):
    # Tokenizer will automatically set [BOS] <text> [EOS]
    # cut off at BERT max length 512
    inputs = tokenizer(batch["article"], padding="max_length", truncation=True, max_length=512, return_tensors="pt")
    input_ids = inputs.input_ids.to('cuda')
    attention_mask = inputs.attention_mask.to('cuda')

    with autocast():
        outputs = bert2bertMK.generate(input_ids, attention_mask=attention_mask)

    # all special tokens including will be removed
    output_str = tokenizer.batch_decode(outputs, skip_special_tokens=True)

    batch["pred"] = output_str

    return batch

results = test_data.map(generate_summary, batched=True, batch_size=batch_size, remove_columns=["article"])

pred_str = results["pred"]
label_str = results["highlights"]

rouge_output = rouge.compute(predictions=pred_str, references=label_str, rouge_types=["rouge2"])["rouge2"].mid

print(rouge_output)

Map:   0%|          | 0/5700 [00:00<?, ? examples/s]

Score(precision=0.15225139568423074, recall=0.1674391018626793, fmeasure=0.15436972362742502)


### Show some sample summaries with the fine-tuned model

**Function to decode tensors**

In [ ]:
def tensor_to_text(tensor_ids):
    # Convert tensor IDs to text
    text = tokenizer.decode(tensor_ids, skip_special_tokens=True)
    return text

In [ ]:
def summarize_article(article, tokenizer, summarizer):
    inputs = tokenizer(article, return_tensors="pt", max_length=512, truncation=True)
    inputs = {key: val.to("cuda") for key, val in inputs.items()}  # Ensure inputs are on GPU
    summary_ids = summarizer.generate(
        inputs['input_ids'],
        attention_mask=inputs['attention_mask'],
        max_length=150,  # Adjust based on your summarization length
        num_beams=4,
        length_penalty=2.0,
        early_stopping=True,
        do_sample=True  # Set to True for more diverse summaries
    )
    summary = tokenizer.decode(summary_ids[0], skip_special_tokens=True)
    return summary

**Generating a few sample summaries with it**

In [ ]:
news_articles = [tensor_to_text(example["input_ids"]) for example in train_data.select(range(3))]

Summarize the articles and store summaries in summaries list. We set `do_sample` to `True` to avoid greedy decoding and provide a more natural, human-like summary.

In [ ]:
fine_tuned_summaries = []
for article in news_articles:
    try:
        summary = summarize_article(article, tokenizer, bert2bertMK)
        fine_tuned_summaries.append(summary)
    except Exception as e:
        fine_tuned_summaries.append("Error, article was too long")

Create a dataframe containing the fine tuned BERT summaries and the target summaries from our fine-tuning dataset

In [ ]:
summaries = pd.DataFrame({
    'Fine-tuned bert2BERTMK  Summaries': fine_tuned_summaries,
    'Target Summaries': target_summaries
})

In [ ]:
fine_tuned_summaries

['Harry Potter star Daniel Radcliffe earns $ 41. 1 million fortune. The 18 - year - old will be able to gamble in a casino or see the horror film " Hostel : Part II " Radcliffe\'s earnings from the first five Potter films have been held in a trust fund. He\'ll also appear in " December Boys " and " December Boy "',
 'Miami - Dade pretrial detention facility dubbed " the forgotten floor " Judge Steven Leifman says inmates with the most severe mental illnesses are incarcerated. Inmate : " I am the son of the president of the presidency " Prosecutors say one - third of inmates are mentally ill.',
 'NEW : " I probably had a 30 -, 35 - foot free fall, " survivor says. " I could see the whole bridge as it was going down, " a passenger says. Rescue efforts were controlled and organized, emergency room physician says. The Minnesota bridge fell all the way down, hitting a broken - off section.']

In [ ]:
target_summaries

["Harry Potter star Daniel Radcliffe gets £20M fortune as he turns 18 Monday .\nYoung actor says he has no plans to fritter his cash away .\nRadcliffe's earnings from first five Potter films have been held in trust fund .",
 'Mentally ill inmates in Miami are housed on the "forgotten floor"\nJudge Steven Leifman says most are there as a result of "avoidable felonies"\nWhile CNN tours facility, patient shouts: "I am the son of the president"\nLeifman says the system is unjust and he\'s fighting for change .',
 'NEW: "I thought I was going to die," driver says .\nMan says pickup truck was folded in half; he just has cut on face .\nDriver: "I probably had a 30-, 35-foot free fall"\nMinnesota bridge collapsed during rush hour Wednesday .']

In [ ]:
summaries

Fine-tuned bert2BERTMK  Summaries  \
0  Harry Potter star Daniel Radcliffe earns $ 41. 1 million fortune. The 18 - year - old will be able to gamble in a casino or see the horror film " Hostel : Part II " Radcliffe's earnings from the first five Potter films have been held in a trust fund. He'll also appear in " December Boys " and " December Boy "   
1                                 Miami - Dade pretrial detention facility dubbed " the forgotten floor " Judge Steven Leifman says inmates with the most severe mental illnesses are incarcerated. Inmate : " I am the son of the president of the presidency " Prosecutors say one - third of inmates are mentally ill.   
2                  NEW : " I probably had a 30 -, 35 - foot free fall, " survivor says. " I could see the whole bridge as it was going down, " a passenger says. Rescue efforts were controlled and organized, emergency room physician says. The Minnesota bridge fell all the way down, hitting a broken - off section.   

                                                                                                                                                                                                                                                                               Target Summaries  
0                                                                   Harry Potter star Daniel Radcliffe gets £20M fortune as he turns 18 Monday .\nYoung actor says he has no plans to fritter his cash away .\nRadcliffe's earnings from first five Potter films have been held in trust fund .  
1  Mentally ill inmates in Miami are housed on the "forgotten floor"\nJudge Steven Leifman says most are there as a result of "avoidable felonies"\nWhile CNN tours facility, patient shouts: "I am the son of the president"\nLeifman says the system is unjust and he's fighting for change .  
2                                                           NEW: "I thought I was going to die," driver says .\nMan says pickup truck was folded in half; he just has cut on face .\nDriver: "I probably had a 30-, 35-foot free fall"\nMinnesota bridge collapsed during rush hour Wednesday .

We see that our model performs great on summaries with a Rouge-2 score of **15.5.** which is somewhat close to what Patrick Von Platen reports on hugging face **18.22**.

# Task 4
Save the fine-tuned model

###  Save Fine-Tuned Model
Save the model and tokenizer to the 'model' folder in the current directory

In [ ]:
bert2bert.save_pretrained('./model')

### Save Tokenizer

In [ ]:
tokenizer.save_pretrained('./model')

('./model/tokenizer_config.json',
 './model/special_tokens_map.json',
 './model/vocab.txt',
 './model/added_tokens.json',
 './model/tokenizer.json')

### Split model into parts for GitHub upload

In [ ]:
def split_file(file_path, chunk_size_mb=50):
    chunk_size = chunk_size_mb * 1024 * 1024  # Convert MB to bytes
    with open(file_path, 'rb') as f:
        chunk = f.read(chunk_size)
        i = 0
        while chunk:
            with open(f"{file_path}.part{i}", 'wb') as chunk_file:
                chunk_file.write(chunk)
            i += 1
            chunk = f.read(chunk_size)

In [ ]:
model_dir = './model'

In [ ]:
# Split only the .bin file
for filename in os.listdir(model_dir):
    if filename.endswith('.bin'):  # Only split .bin files
        file_path = os.path.join(model_dir, filename)
        split_file(file_path, chunk_size_mb=50)  # Split into 50 MB chunks

In [ ]:
zip_file_name = './model.zip'
shutil.make_archive(zip_file_name.replace('.zip', ''), 'zip', model_dir)

print(f"Model saved and zipped at {zip_file_name}")

Model saved and zipped at ./model.zip


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Now our model is trained and saved we can use it in our deployed application to summarize news articles in real-time!

# References

**Bibliography**  
[1]Snowflake Inc. 2024. Connect Streamlit to Google Cloud Storage - Streamlit Docs. docs.streamlit.io. Retrieved July 19, 2024 from https://docs.streamlit.io/develop/tutorials/databases/gcs  
[2]Keras Team. 2024. Keras documentation: InceptionV3. keras.io. Retrieved July 19, 2024 from https://keras.io/api/applications/inceptionv3/  
[3]TensorFlow. 2024. Load video data | TensorFlow Core. TensorFlow. Retrieved July 19, 2024 from https://www.tensorflow.org/tutorials/load_data/video#create_frames_from_each_video_file  
[4]TensorFlow. 2024. tf.keras.applications.inception_v3.decode_predictions | TensorFlow v2.16.1. TensorFlow. Retrieved July 19, 2024 from https://www.tensorflow.org/api_docs/python/tf/keras/applications/inception_v3/decode_predictions

# Phase 5         
Deploy Functional App

### Natural Language Processing

This notebook addresses Phase 4 of the What's The Hot Topic in Topic Town? project: **Deploy Functional App**.  
The self-created rubric, in our repository, explains the requirement for a proper execution of this phase as seen below.   
**Description:** Use Streamlit to build an interactive and responsive web application for project functionality

**Note:** This notebook **only** shows the code and instructions for deployment, specifically writing the Streamlit app. As such, you cannot run this notebook. You must have already build your Spider with scrapy before attempting to write a Streamlit app as the Streamlit  integrates all the elements of **Phase 1-5** in one working system. If you have not built your Spider yet, you can learn how to build one using this link: [Scrapy Tutorial](https://docs.scrapy.org/en/latest/intro/tutorial.html)

### Repository Link

Here is a link to our repository:

[What's The Hot Topic In Town?](https://github.com/kelvin-ahiakpor/Whats.The.Hot.Topic.In.Town)

### Imports

In [ ]:
import phase2_phase3 as news
import streamlit as st
import sqlite3
import pandas as pd
import subprocess
import os
import sys
import tempfile
from pathlib import Path
from transformers import TFAutoModelForSeq2SeqLM, AutoTokenizer
import io

### Function to reassmble split model files

In [ ]:
def reassemble_file(chunk_prefix, output_file_path, input_dir='model'):
    chunk_files = sorted([f for f in os.listdir(input_dir) if f.startswith(chunk_prefix)])
    with open(output_file_path, 'wb') as output_file:
        for chunk_file_name in chunk_files:
            with open(os.path.join(input_dir, chunk_file_name), 'rb') as chunk_file:
                output_file.write(chunk_file.read())

### Function to load the model

In [ ]:
# Function to load the model from the combined bytes
def load_model():
    output_file_path = '../bert2bertMK/model.safetensors'  # Path for the reassembled model
    chunk_prefix = 'model.safetensors_chunk_'
    reassemble_file(chunk_prefix, output_file_path, input_dir='../bert2bertMK/model')

    # Load the model and tokenizer
    tokenizer = AutoTokenizer.from_pretrained('../bert2bertMK')
    model = TFAutoModelForSeq2SeqLM.from_pretrained('../bert2bertMK')

    return model, tokenizer

### Function to summarize an article

In [ ]:
def summarize_article(article_text, model, tokenizer):
    if model is None or tokenizer is None:
        return "Error: Model or tokenizer not loaded properly."
    
    try:
        inputs = tokenizer.encode("summarize: " + article_text, return_tensors="tf", max_length=512, truncation=True)
        outputs = model.generate(inputs, max_length=150, min_length=40, length_penalty=2.0, num_beams=4, early_stopping=True)
        summary = tokenizer.decode(outputs[0], skip_special_tokens=True)
        return summary
    except Exception as e:
        return f"Error during summarization: {str(e)}"

## Streamlit app

In [ ]:
menu = ["HOME", "SEARCH"]
choice = st.sidebar.selectbox("Menu", menu)

st.sidebar.title("About")
st.sidebar.info(
    """
     This app informs users about the hottest trending topics in their country and provides sentiment analysis and concise summaries. Users can access the original articles as well.
     
    GitHub: [What'sTheHotTopicInTown?](https://github.com/kelvin-ahiakpor/Whats.The.Hot.Topic.In.Town)
    """
)

if choice ==  "HOME":
    # Main title
    st.title("Welcome to TrendWatch!")

    # Catchy writeup
    st.markdown("""
    ### Experience The Future of News! 
    Your go-to app for discovering the hottest trending topics in your country. Our cutting-edge technology ensures you're always in the loop with what's happening around you.

    **Why Choose TrendWatch?**

    - **Instant Updates**: Get the latest trending topics as they happen, scraped in real-time using the powerful Scrapy library.
    - **Deep Insights**: Understand the public mood with our advanced sentiment analysis powered by the state-of-the-art Distill-BERT model.
    - **Quick Reads**: No time to read long articles? Our Bert2BertMK model generates concise summaries, so you get the gist in no time.
    - **Stay Informed**: Dive deeper into topics that interest you with direct links to the full articles.

    **Stay Ahead of The Curve**

    With TrendWatch, you'll never miss out on the trends that matter. Whether it's breaking news, viral topics, or the latest buzz, we've got you covered. Our sleek and user-friendly interface ensures you get all the information you need at a glance.

    Stay up to date, more details on our community will be provided soon.
    """)

    # Contact information
    st.markdown("""
    **Contact Us:**
    - Emmanuel Acquaye: [emmanuel.acquaye@ashesi.edu.gh](mailto:emmanuel.acquaye@ashesi.edu.gh)
    - Kelvin Ahiakpor: [kelvin.ahiakpor@ashesi.edu.gh](mailto:kelvin.ahiakpor@ashesi.edu.gh)
    - **Student IDs**: 10112026, 47822026
    - **GitHub**: [WHAT'S THE HOT TOPIC IN TOWN?](https://github.com/kelvin-ahiakpor/Whats.The.Hot.Topic.In.Town)
    """)

    # Start Searching Now button
    if st.button('Start Searching Now'):
        st.session_state.page = 'search'
        
elif choice == "SEARCH":
    # Main title
    st.title("Search for Trending Topics")

    with st.expander("HOW TO USE?"):
        st.markdown(
            """
            **INSTRUCTIONS:**
            - Type in the African Country of choice and click the Get Trending News Button.
            - The topics are then displayed to you with their summaries and sentiments
            
            **Contact Information:**
            - Email: emmanuel.acquaye@ashesi.edu.gh, kelvin.ahiakpor@ashesi.edu.gh
            - Student ID: 10112026, 47822026
            
            - GitHub: [WHAT'S THE HOT TRENDING TOPIC IN TOWN](https://github.com/kelvin-ahiakpor/Whats.The.Hot.Topic.In.Town)
            """
        )

    top_articles = None

    def run_scrapy_script(country):
        result = subprocess.run([sys.executable, "RunSpider.py", country], capture_output=True, text=True)
        print("scraped")
        return result.stdout

    st.subheader("FIND THE TOP TRENDS OF AFRICAN COUNTRIES")
    country = st.text_input("Enter Country For News")
    scrape_button = st.button("Get Trending News")

    if scrape_button and country:
        with st.spinner('Scraping data...'):
            output = run_scrapy_script(country)
        st.success("Scraping completed!")
        
        conn = sqlite3.connect('newsData.db')
        newsdata = news.load_and_clean_data(conn)
        top_terms = news.get_top_terms(newsdata)
        top_articles = news.calculate_relevance(newsdata, top_terms)
        model, tokenizer = load_model()

        if model is None or tokenizer is None:
            st.error("Failed to load the model or tokenizer. Please check the console for more information.")
        elif top_articles is not None and not top_articles.empty:
            st.subheader("TOP 10 TRENDING TOPICS")
            
            article_titles = top_articles['TITLE'].tolist()
            
            for i in range(len(article_titles)):
                st.subheader(str(i+1) + ". " + article_titles[i])
                article_content = top_articles[top_articles['TITLE'] == article_titles[i]]['BODY'].values[0]
                summary = summarize_article(article_content, model, tokenizer)
                st.write(summary)
                st.write()
                link = top_articles[top_articles['TITLE'] == article_titles[i]]['URL'].values[0]
                sentiment = news.analyze_sentiment(article_content)
                if sentiment == "POSITIVE":
                    st.write("Sentiment Analysis:  😊 ", sentiment)
                else:
                    st.write("Sentiment Analysis: 😡 ", sentiment)
                    
                st.write("Link To Article: ", link)
        else:
            st.write("No articles found or error in processing data.")

**Thanks for reading**